### PR TITLE
update translation message

### PR DESCRIPTION
**Pull request by DarjaLeonova**

### CODE REVIEW

File: cmd/translate/main_grpc_integration_test.go [hunk 1]
```diff
@@ -684,33 +686,43 @@ func Test_UpdateTranslation_gRPC(t *testing.T) {
...
 		expectedCode codes.Code
 	}{
 		{
-			name:         "Happy Path",
+			name:         "Happy Path update all",
 			request:      happyReq,
 			expectedCode: codes.OK,
 		},
+		{
+			name:         "Happy path update original field",
```
Comment (klavinsjanis) [inline cmd/translate/main_grpc_integration_test.go:722]: update_mask 'original'? What about update_mask targeting 'messages' field?

File: pkg/model/mask.go [hunk 2]
```diff
@@ -135,3 +157,17 @@ func UpdateService(src, dst *Service, mask Mask) error {
...
 }
+
+// UpdateTranslation updates the destination translation based on the source translation and field mask.
+func UpdateTranslation(src, dst *Translation, mask Mask) error {
+	// Prevent updating read-only fields like ID
+	if slices.ContainsFunc(mask, func(s string) bool {
+		return strings.EqualFold(s, "ID")
+	}) {
+		return errors.New("\"id\" is not allowed in field mask")
+	}
```
Comment (VladislavsPerkanuks) [inline pkg/model/mask.go:168]: Translation does not have any read-only fields, and also it does not have ID.
so possible values for mask is
```go
language
messages
original
```

File: pkg/model/mask.go [hunk 3]
```diff
@@ -115,6 +110,33 @@ func updateField(src, dst reflect.Value, fields []string) {
 	}
 }
 
+// updateSliceField this function updates a destination slice by either replacing a struct with a matching ID
+// or appending new values if no matching ID is found.
+func updateSliceField(srcField, dstField reflect.Value) {
```
Comment (klavinsjanis) [inline pkg/model/mask.go:115]: Maybe we should consider using a custom implementation for update_mask 'messages' since it might be used frequently.

File: pkg/model/mask.go [hunk 4]
```diff
@@ -135,3 +157,17 @@ func UpdateService(src, dst *Service, mask Mask) error {
 
 	return nil
 }
+
+// UpdateTranslation updates the destination translation based on the source translation and field mask.
+func UpdateTranslation(src, dst *Translation, mask Mask) error {
```
Comment (VladislavsPerkanuks) [inline pkg/model/mask.go:162]: I would go with following logic, that would simplify everything

1. Check mask, if it does not contain messages, then just call update()
2. If it exists, then implement logic of updating slice here, not in update() function. Pop the messages from mask, and call update, if mask is not empty yet

File: cmd/translate/main_grpc_integration_test.go (hunk 1 above)
Comment (VladislavsPerkanuks) [reply cmd/translate/main_grpc_integration_test.go:722]: does not matter, as this test do not check, if implementation doing the right thing, it just checks that request gone through and returned ok code.

tho It could test, not only the status code, but also the returned value

File: cmd/translate/main_rest_integration_test.go [hunk 5]
```diff
@@ -640,22 +640,22 @@ func Test_UpdateTranslation_REST(t *testing.T) {
 	createTranslation(ctx, t, service.GetId(), &translatev1.Translation{Original: true, Language: langs[0].String()})
 	createTranslation(ctx, t, service.GetId(), &translatev1.Translation{Original: false, Language: langs[1].String()})
 
-	happyReq := randUpdateTranslationReq(t, service.GetId(), &translatev1.Translation{Language: langs[1].String()})
+	happyReq := randUpdateTranslationReq(t, service.GetId(), &translatev1.Translation{Language: langs[1].String()}, nil)
```
Comment (VladislavsPerkanuks) [inline cmd/translate/main_rest_integration_test.go:643]: missing test case with field mask for REST

File: pkg/model/mask_test.go [hunk 6]
```diff
@@ -299,3 +301,92 @@ func Test_UpdateServiceFromMask(t *testing.T) {
...
+				require.Equal(t, original.Language, dstTranslation.Language)
+				require.Equal(t, original.Original, dstTranslation.Original)
+			},
+		},
+		{
+			name:        "Try to update ID",
+			fieldMask:   Mask{"ID"},
+			expectedErr: errors.New("\"id\" is not allowed in field mask"),
+		},
+	}
```
Comment (VladislavsPerkanuks) [inline pkg/model/mask_test.go:372]: there is no id in tranlsation resource

File: pkg/server/translation.go [hunk 7]
```diff
@@ -248,6 +253,12 @@ func (t *TranslateServiceServer) UpdateTranslation(
 		}
 	}
 
+	for i := range all {
+		if err = model.UpdateTranslation(params.translation, &all[i], params.mask); err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, err.Error())
+		}
+	}
```
Comment (VladislavsPerkanuks) [inline pkg/server/translation.go:260]: That will break things.

Scenario
```go
// in DB
// Translation1 with Language LV
// Translation2 with Language En

// Request to update Translation1 language with new language RU

// The result of this selected code should be
// Translation1 with Language RU
// Translation2 with Language RU

// And then saving to repo, will try to save multiple translation into one composite key (service + language)
```

File: pkg/model/mask.go (hunk 3 above)
Comment (janishorsts) [reply pkg/model/mask.go:115]: When we have a slice with a struct element with `ID` field, we assume it is the unique identifier and we can update this element.

What update mask do we use to update the Message in Translation? `messages.*`?

Comment (janishorsts) [reply pkg/model/mask.go:115]: https://github.com/protocolbuffers/protobuf/blob/9bbea4aa65bdaf5fc6c2583e045c07ff37ffb0e7/src/google/protobuf/field_mask.proto#L111

```proto
// If a repeated field is specified for an update operation, new values will
// be appended to the existing repeated field in the target resource. Note that
// a repeated field is only allowed in the last position of a `paths` string.
```

I assume, that when the update mask is a repeated field, we append by default. However, if the repeated message contains the field `id` (unique identifier), we replace it, right?

In our case, update mask `messages` replaces the message by its id:
* appends if it is a new message
* replaces if it is an existing message

File: pkg/model/mask.go (hunk 4 above)
Comment (janishorsts) [reply pkg/model/mask.go:162]: We should ignore managing read-only fields for now and determine general-purpose and reusable solutions later.

It should be closer to the request validation. Events (gRPC etc) that have reached inside our system MUST be already safe to process.

```go
// allowMask returns true if updateMask does not contain any readOnlyMask.
func allowMask(updateMask, readOnlyMask Mask) bool
```

File: pkg/model/mask.go (hunk 3 above)
Comment (DarjaLeonova) [reply pkg/model/mask.go:115]: @janishorsts Currently the updatesSliceField function appends a new message if ID doesn't exist and replaces a message if ID exists. You can check tests in mask_test.go. So as I understood it is expected behavior, right?

File: pkg/model/mask.go (hunk 4 above)
Comment (VladislavsPerkanuks) [reply pkg/model/mask.go:165]: there is no mention of read-only fields in my comment

Comment (janishorsts) [reply pkg/model/mask.go:165]: We don't need to define `func Update<resource>` for each resource. We can get away with general-purpose functions.

File: pkg/model/mask.go [hunk 8]
```diff
@@ -135,3 +160,43 @@ func UpdateService(src, dst *Service, mask Mask) error {
...
+	// nonMsgField will store the field paths that are not related to "Messages."
+	nonMsgField := []string{}
+
+	for _, path := range mask {
+		switch path {
+		case "Messages":
```
Comment (janishorsts) [inline pkg/model/mask.go:171]: This works only for Translation.

How can you implement this as a general purpose solution in `updateField`? Based on https://github.com/translate-agent/translate/pull/219/files#r1437547444

`when the update mask is a repeated field, we append by default. However, if the repeated message contains the field id (unique identifier), we replace it`

File: pkg/model/mask.go [hunk 9]
```diff
@@ -115,6 +110,50 @@ func updateField(src, dst reflect.Value, fields []string) {
...
+		_, ok := el.FieldByName("ID")
+
+		return !ok
+	}
+
+	if appendOnly() {
+		// If the field is a slice, append new values from src to existing slice in dst
+		//nolint:lll
+		// https://github.com/protocolbuffers/protobuf/blob/9bbea4aa65bdaf5fc6c2583e045c07ff37ffb0e7/src/google/protobuf/field_mask.proto#L111
+		dstField.Set(reflect.AppendSlice(dstField, srcField))
```
Comment (janishorsts) [inline pkg/model/mask.go:132]: We can remove this comment but move the link to protobuf description to func description.
```suggestion
		dstField.Set(reflect.AppendSlice(dstField, srcField))
```

File: pkg/model/mask.go [hunk 10]
```diff
@@ -115,6 +110,50 @@ func updateField(src, dst reflect.Value, fields []string) {
...
+				found = true
+			}
+		}
+
+		if !found {
+			dstField.Set(reflect.Append(dstField, srcField.Index(i)))
+			continue
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (VladislavsPerkanuks) [reply pkg/model/mask.go:155]: i would go with
```go

func findInDstField(srcID interface{}, dstField reflect.Value) (int, bool) {
	for i := 0; i < dstField.Len(); i++ {
			if dstField.Index(j).FieldByName("ID").Interface() == srcID {
					return i
			}
	}
	return -1
}

...

srcID := srcField.Index(i).FieldByName("ID").Interface()
idx := findInDstField(srcID, dstField)

// idx == -1 -> append
// idx != -1 -> replace

```

File: pkg/model/mask.go [hunk 11]
```diff
@@ -115,6 +110,50 @@ func updateField(src, dst reflect.Value, fields []string) {
 	}
 }
 
+// updateSliceField this function updates a destination slice by either replacing a struct with a matching ID
+// or appending new values if no matching ID is found.
+func updateSliceField(srcField, dstField reflect.Value) {
+	appendOnly := func() bool {
```
Comment (VladislavsPerkanuks) [inline pkg/model/mask.go:116]: A litle more simple
```suggestion
appendOnly := func() bool {
    el := srcField.Type().Elem()
    _, ok := el.FieldByName("ID")
     
     return el.Kind() != reflect.Struct || !ok
    }
```

File: pkg/model/mask.go [hunk 12]
```diff
@@ -115,6 +110,50 @@ func updateField(src, dst reflect.Value, fields []string) {
...
+		el := srcField.Type().Elem()
+
+		if el.Kind() != reflect.Struct {
+			return true
+		}
+
+		_, ok := el.FieldByName("ID")
+
+		return !ok
```
Comment (VladislavsPerkanuks) [inline pkg/model/mask.go:125]: A litle more simple
```suggestion
appendOnly := func() bool {
    el := srcField.Type().Elem()
    _, ok := el.FieldByName("ID")
     
     return el.Kind() != reflect.Struct || !ok
    }
```

File: pkg/model/mask.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (janishorsts) [reply pkg/model/mask.go:155]: https://pkg.go.dev/slices#IndexFunc

File: pkg/model/mask.go (hunk 4 above)
Comment (janishorsts) [reply pkg/model/mask.go:179]: We don't need `UpdateTranslation` anymore.

Let's expose `update` and use it instead.

File: proto/translate/v1/translate.proto [hunk 13]
```diff
@@ -83,6 +83,7 @@ message UpdateTranslationRequest {
   string service_id = 1;
   bool populate_translations = 2;
   Translation translation = 3;
+  google.protobuf.FieldMask update_mask = 4;
 }
```
Comment (janishorsts) [inline proto/translate/v1/translate.proto:87]: Let's change the order of fields in the proto message to make it more logical order.

PS. This cannot be done when the code is in the production.

```suggestion
  string service_id = 1;
  Translation translation = 2;
  google.protobuf.FieldMask update_mask = 3;
  bool populate_translations = 4;
}
```

File: cmd/translate/main_rest_integration_test.go [hunk 14]
```diff
@@ -640,22 +641,32 @@ func Test_UpdateTranslation_REST(t *testing.T) {
...
+
+	req := randUpdateTranslationReq(t, service.GetId(), &translatev1.Translation{
+		Language: langs[1].String(),
+		Messages: []*translatev1.Message{
+			{
+				Id: "Hello", Message: "World",
+			},
+		},
+	},
+		&field_mask.FieldMask{Paths: []string{"messages"}})
```
Comment (klavinsjanis) [inline cmd/translate/main_rest_integration_test.go:654]: Maybe could add same test to grpc client integration tests as well.

File: pkg/model/mask.go (hunk 12 above)
Comment (DarjaLeonova) [reply pkg/model/mask.go:125]: This solution will cause a panic. We should ensure that the type on which we are calling FieldByName is indeed a struct type.

Comment (VladislavsPerkanuks) [reply pkg/model/mask.go:125]: oki

File: pkg/server/translation.go [hunk 15]
```diff
@@ -248,6 +253,13 @@ func (t *TranslateServiceServer) UpdateTranslation(
 		}
 	}
 
+	if err = model.UpdateTranslation(
```
Comment (klavinsjanis) [inline pkg/server/translation.go:256]: Could you add an integration test where a message for an original translation is updated/appended ? 
So that we know everything is working as intended.

**Review by VladislavsPerkanuks — APPROVED** [review]
From my perspective looks fine. Fix other people's comments and should be good.

And let's hope it does not bring any unintended behavior into the system 🙈

File: pkg/model/mask.go [hunk 16]
```diff
@@ -115,6 +110,53 @@ func updateField(src, dst reflect.Value, fields []string) {
...
+			// If found, update the corresponding structure in the destination slice
+			dstField.Index(idx).Set(srcField.Index(i))
+			continue
+		}
+
+		dstField.Set(reflect.Append(dstField, srcField.Index(i)))
+	}
+}
+
+func findInDstField(srcID interface{}, dstField reflect.Value) int {
```
Comment (janishorsts) [inline pkg/model/mask.go:150]: `func updateSliceField` is relatively simple.

Can we add this the same way you did with `appendOnly` to avoid package level functions?

Simply `index := func(id any) int {}`?

File: pkg/model/mask.go (hunk 12 above)
Comment (janishorsts) [reply pkg/model/mask.go:125]: It is less performant, too.

Comment (janishorsts) [reply pkg/model/mask.go:125]: @VladislavsPerkanuks I like the thinking, though. But check my previous comment. :)

File: proto/translate/v1/translate.proto [hunk 17]
```diff
@@ -150,6 +151,8 @@ service TranslateService {
     };
   }
 
+  //nolint:lll
```
Comment (janishorsts) [inline proto/translate/v1/translate.proto:154]: `.proto` is not golang source code, isn't it? `//nolint:` is applicable only the `.go` files.

File: proto/translate/v1/translate.proto [hunk 18]
```diff
@@ -150,6 +151,8 @@ service TranslateService {
...
+  // https://github.com/protocolbuffers/protobuf/blob/9bbea4aa65bdaf5fc6c2583e045c07ff37ffb0e7/src/google/protobuf/field_mask.proto#L111
```
Comment (janishorsts) [inline proto/translate/v1/translate.proto:155]: Why did you add this comment here?

File: pkg/model/mask.go [hunk 19]
```diff
@@ -115,6 +110,53 @@ func updateField(src, dst reflect.Value, fields []string) {
...
+	// If found by id, replace the corresponding structure in the destination slice. Otherwise, append.
+	for i := 0; i < srcField.Len(); i++ {
+		srcID := srcField.Index(i).FieldByName("ID").Interface()
+		idx := findInDstField(srcID, dstField)
+
+		// Check if the "ID" field of the first element in the source slice
+		// matches the "ID" field of the current element in the destination slice.
+		if idx != -1 {
+			// If found, update the corresponding structure in the destination slice
+			dstField.Index(idx).Set(srcField.Index(i))
```
Comment (janishorsts) [inline pkg/model/mask.go:142]: The code is self-descriptive. No need for comments.

My personal preference, emphasizes all valid values, inclusive range.

```suggestion
		if idx >= 0 {
			dstField.Index(idx).Set(srcField.Index(i))
```
Emphasizes all valid values, exclusive range.

```suggestion
		if idx > -1 {
			dstField.Index(idx).Set(srcField.Index(i))
```

**Review by klavinsjanis — APPROVED** [review]

**Comment by klavinsjanis** [comment]
Did you merge in origin/main?

File: cmd/translate/main_grpc_integration_test.go [hunk 20]
```diff
@@ -82,7 +84,7 @@ func createService(ctx context.Context, t *testing.T) *translatev1.Service {
 // createTranslation creates a random translation, and calls the CreateTranslation RPC.
 func createTranslation(ctx context.Context, t *testing.T, serviceID string,
 	override *translatev1.Translation,
-) *translatev1.Translation { //nolint:unparam
+) *translatev1.Translation { //nolintlint:unparam
```
Comment (klavinsjanis) [inline cmd/translate/main_grpc_integration_test.go:87]: is this nolint needed?

Comment (DarjaLeonova) [reply cmd/translate/main_grpc_integration_test.go:87]: Looks like //nolint:unparam is not needed at all ;D

File: cmd/translate/main_grpc_integration_test.go [hunk 21]
```diff
@@ -672,6 +674,272 @@ func Test_ListTranslations_gRPC(t *testing.T) {
...
+	service7 := createService(ctx, t)
+	service8 := createService(ctx, t)
+
+	tests := []struct {
+		request     *translatev1.UpdateTranslationRequest
+		translation *translatev1.Translation
+		name        string
+	}{
+		{
+			name: "Update messages field in original translation",
```
Comment (klavinsjanis) [inline cmd/translate/main_grpc_integration_test.go:716]: Could you group these tests? 
Right now they are a. bit mixed up.
Something like -
update message/s for translation, 
update message/s for original, 
update original status for translation

File: cmd/translate/main_grpc_integration_test.go [hunk 22]
```diff
@@ -672,6 +674,272 @@ func Test_ListTranslations_gRPC(t *testing.T) {
...
+		tt := tt
+		subtest(tt.name, func(ctx context.Context, t *testing.T) {
+			beforeUpdateTranslations, err := client.ListTranslations(ctx, &translatev1.ListTranslationsRequest{
+				ServiceId: tt.request.GetServiceId(),
+			})
+
+			require.NoError(t, err)
+			resp, err := client.UpdateTranslation(ctx, tt.request)
+
+			if err == nil {
```
Comment (klavinsjanis) [inline cmd/translate/main_grpc_integration_test.go:914]: why do you check if err is nil? Isn't all of these happy path tests?

File: cmd/translate/main_grpc_integration_test.go [hunk 23]
```diff
@@ -672,6 +674,272 @@ func Test_ListTranslations_gRPC(t *testing.T) {
...
+			afterUpdateTranslations, err := client.ListTranslations(ctx, &translatev1.ListTranslationsRequest{
+				ServiceId: tt.request.GetServiceId(),
+			})
+
+			require.NoError(t, err)
+			require.NotEmpty(t, resp)
+
+			var translationFromService *translatev1.Translation
+
+			for i := range afterUpdateTranslations.GetTranslations() {
```
Comment (janishorsts) [inline cmd/translate/main_grpc_integration_test.go:927]: Never let actual values guide your tests. Use your expectations to guide.

E.g. `ListTranslations` returns an empty list, the test passes, right? But, actually, it should fail.

File: cmd/translate/main_grpc_integration_test.go [hunk 24]
```diff
@@ -672,6 +674,272 @@ func Test_ListTranslations_gRPC(t *testing.T) {
...
+			request: randUpdateTranslationReq(t, service4.GetId(), &translatev1.Translation{
+				Language: langs[0].String(),
+				Messages: []*translatev1.Message{
+					{
+						Id: "Hello", Message: "World",
+					},
+				},
+			},
+				&field_mask.FieldMask{Paths: []string{"messages"}}),
+			translation: createTranslation(ctx, t, service4.GetId(), &translatev1.Translation{
```
Comment (janishorsts) [inline cmd/translate/main_grpc_integration_test.go:789]: All tests MUST be isolated - no resource sharing. Any reason why you use the same service as in the previous test?

Comment (DarjaLeonova) [reply cmd/translate/main_grpc_integration_test.go:789]: I will fix it. 👍

File: pkg/model/mask.go [hunk 25]
```diff
@@ -54,7 +54,7 @@ type Mask []string
 //
 //	update(&src, &dst, mask)
```
Comment (mvilks) [inline pkg/model/mask.go:55]: Uppercase

**Review by mvilks — APPROVED** [review]

File: pkg/model/mask.go [hunk 26]
```diff
@@ -115,6 +110,51 @@ func updateField(src, dst reflect.Value, fields []string) {
...
+		}
+
+		index := idx(srcID)
+
+		if index >= 0 {
+			dstField.Index(index).Set(srcField.Index(i))
+			continue
+		}
+
+		dstField.Set(reflect.Append(dstField, srcField.Index(i)))
```
Comment (VladislavsPerkanuks) [inline pkg/model/mask.go:154]:
```suggestion

		switch j := idx(srcID); j {
		default: // found -> replace
			dstField.Index(j).Set(srcField.Index(i))
		case -1: // not found -> append
			dstField.Set(reflect.Append(dstField, srcField.Index(i)))
		}
```

**Review by VladislavsPerkanuks — APPROVED** [review]

File: pkg/model/mask.go (hunk 25 above)
Comment (janishorsts) [reply pkg/model/mask.go:55]:
```suggestion
//	Update(&src, &dst, mask)
```

File: pkg/model/mask.go [hunk 27]
```diff
@@ -33,7 +33,7 @@ import (
 //	...
 type Mask []string
 
-// update updates the dst with the values from src based on the mask.
+// Update updates the dst with the values from src based on the mask.
```
Comment (janishorsts) [inline pkg/model/mask.go:36]: Please describe how slices are handled in the exposed func.

```go
//
// When updating slices:
//   - replace the element if slice contains a struct with a field "ID"
//   - append in all other cases
```
